spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/TensorflowToBigDL.scala

```diff
@@ -0,0 +1,1091 @@
+import scala.reflect.{ClassTag, classTag}
+
+/**
+ * Represent a mapping from tensorflow operations graph to BigDL Module
+ */
+trait TensorflowToBigDL {
+
+  /**
+   * The topology of the tensorflow operation graph
+   * @return
+   */
+  def topology: DirectedGraph[String]
+
+  /**
+   * Get the BigDL model
+   * @param tfGraph operation graph
+   * @param context variables
+   * @return (module, input nodes, output nodes)
+   */
+  def layer[T: ClassTag](
+    tfGraph: DirectedGraph[NodeDef],
+    context: Context[T],
+    byteOrder: ByteOrder,
+    dataFormat: TensorflowDataFormat
+  )(implicit ev: TensorNumeric[T]): AbstractModule[Activity, Tensor[T], T]
+
+  protected def getOrSetTensor[T: ClassTag](
+    node: NodeDef, context: Context[T], byteOrder: ByteOrder)(f: Tensor[T] => Tensor[T])(
+    implicit ev: TensorNumeric[T]): (Tensor[T], Tensor[T]) = {
+
+    if (context.contains(node)) {
+      context(node)
+    } else {
+      val weight = f(toTensor[T](node.getAttrMap.get("value").getTensor, byteOrder)).contiguous()
+      val gradient = Tensor[T](weight.size())
+      context.put(node, (weight, gradient))
+      (weight, gradient)
+    }
+  }
+}
+
+object TensorflowToBigDL {
+
+  /**
+   * Represent one input
+   */
+  val INPUT_PLACEHOLDER: String = "*"
+
+  /**
+   * Represent one or many inputs. Note this can only be the first or the last of the input names
+   */
+  val N_INPUT_PLACEHOLDER: String = "..."
+
+  /**
+   * Separate operation name and its output tensor. In tensorflow, if one operation output multiple
+   * tensors, the tensor will be referred as Op:n, which n is a integer.
+   */
+  val TENSOR_SEPARATOR: String = ":"
+
+  /**
+   * Get the pattern list.
+   * @return
+   */
+  def patterns: Array[TensorflowToBigDL] = {
+    patternList.toArray
+  }
+
+  /**
+   * Register a new mapping from tensor flow operations to BigDL layer. The mapping is defined as
+   * a subclass of TFToBigDL, which defines an operation topology(reversed graph) and how to get
+   * constructor parameters from the topology.
+   * @param pattern
+   */
+  def registerPattern(pattern: TensorflowToBigDL): Unit = {
+    require(pattern.topology.reverse == true, "the topology should be a reversed graph")
+    patternList.append(pattern)
+    sortPattern()
+  }
+
+  var dataFormat: String = "NHWC"
+
+  def dataNCHW: Unit = dataFormat = "NCHW"
+
+  /**
+   * Convert a tensorflow tensor proto to BigDL tensor
+   * @param tfTensor
+   * @return
+   */
+  private[utils] def toTensor[T: ClassTag](tfTensor: TensorProto, endian: ByteOrder)(
+    implicit ev: TensorNumeric[T]): Tensor[T] = {
+
+    require(
+      tfTensor.getDtype == DataType.DT_FLOAT ||
+        tfTensor.getDtype == DataType.DT_FLOAT ||
+        tfTensor.getDtype == DataType.DT_INT32,
+      s"Data type ${tfTensor.getDtype} is not supported now")
+
+    val shape = tfTensor.getTensorShape.getDimList.asScala.map(_.getSize.toInt).toArray
+
+    if (shape.product == 1) {
```

Comment:
Just a thought. Maybe it is not necessary to treat this case specially? If the tensor is a scalar and `shape` is an empty array, we can simply change the shape to Array(1), then the following code will be able to handle the scalar.